internal/controllers/machinedeployment/mdutil/util.go

```diff
@@ -415,45 +416,75 @@ func MachineTemplateDeepCopyRolloutFields(template *clusterv1.MachineTemplateSpe
 // not face a case where there exists a machine set matching the old logic but there does not exist a machineset matching the new logic.
 // In fact previously not matching MS can now start matching the target. Since there could be multiple matches, lets choose the
 // MS with the most replicas so that there is minimum machine churn.
-func FindNewMachineSet(deployment *clusterv1.MachineDeployment, msList []*clusterv1.MachineSet, reconciliationTime *metav1.Time) *clusterv1.MachineSet {
+func FindNewMachineSet(deployment *clusterv1.MachineDeployment, msList []*clusterv1.MachineSet, reconciliationTime *metav1.Time) (*clusterv1.MachineSet, string, error) {
+	if len(msList) == 0 {
+		return nil, "no MachineSets exist for the MachineDeployment", nil
+	}
+
+	// In rare cases, such as after cluster upgrades, Deployment may end up with
+	// having more than one new MachineSets that have the same template,
+	// see https://github.com/kubernetes/kubernetes/issues/40415
+	// We deterministically choose the oldest new MachineSet with matching template hash.
 	sort.Sort(MachineSetsByDecreasingReplicas(msList))
-	for i := range msList {
-		if EqualMachineTemplate(&msList[i].Spec.Template, &deployment.Spec.Template) &&
-			!shouldRolloutAfter(msList[i], reconciliationTime, deployment.Spec.RolloutAfter) {
-			// In rare cases, such as after cluster upgrades, Deployment may end up with
-			// having more than one new MachineSets that have the same template,
-			// see https://github.com/kubernetes/kubernetes/issues/40415
-			// We deterministically choose the oldest new MachineSet with matching template hash.
-			return msList[i]
+
+	var matchingMachineSets []*clusterv1.MachineSet
+	var diffs []string
+	for _, ms := range msList {
+		ms := ms
+
+		equal, diff, err := EqualMachineTemplate(&ms.Spec.Template, &deployment.Spec.Template)
+		if err != nil {
+			return nil, "", errors.Wrapf(err, "failed to compare MachineDeployment spec template with MachineSet %s", ms.Name)
+		}
+		if equal {
+			matchingMachineSets = append(matchingMachineSets, ms)
+		} else {
+			diffs = append(diffs, fmt.Sprintf("MachineSet %s: diff: %s", ms.Name, diff))
 		}
 	}
-	// new MachineSet does not exist.
-	return nil
-}
 
-func shouldRolloutAfter(ms *clusterv1.MachineSet, reconciliationTime *metav1.Time, rolloutAfter *metav1.Time) bool {
-	if ms == nil {
-		return false
+	if len(matchingMachineSets) == 0 {
+		return nil, fmt.Sprintf("couldn't find MachineSet matching MachineDeployment spec template: %s", strings.Join(diffs, ",")), nil
```

Comment:
why did you choose to pass this around instead of logging right here?

Reply:
Because this func is used in multiple places. I wanted to make sure we log those decision only at the point where we actually create a new MachineSet

Reply:
Consdering this one: https://github.com/kubernetes-sigs/cluster-api/pull/10688#pullrequestreview-2084766723

Logging it only when we create a new MachineSet means that we can do it on log level 0.

If we would log it every time we run through this code we have to move it somewhere towards trace level, because it would be very very verbose otherwise. But of course on log levels >=3 it's a lot less useful because nobody will see it per default.

Reply:
Agree overall. I think an error might be not the best fit, because it's not really an error, it's the reasoning why we made a certain decision.